### ISSUE TITLE
NullPointerException if base resource is an empty ResourceCollection

### ISSUE THREAD

**Issue by cowwoc** [eclipse/jetty.project#3278]
Version 9.4.14.v20181114

1. Configure `ServletContextHandler` with `setBaseResource(new ResourceCollection(new Resource[0]))` 
2. `ServletContextHandler.addServlet(DefaultServlet.class, "/")` with `setInitParameter("etags", "true")`
3. Hit the root path on the server
4. Exception will be thrown:

```
java.lang.NullPointerException
	at org.eclipse.jetty.util.resource.Resource.getWeakETag(Resource.java:740)
	at org.eclipse.jetty.util.resource.Resource.getWeakETag(Resource.java:729)
	at org.eclipse.jetty.http.ResourceHttpContent.getETagValue(ResourceHttpContent.java:164)
	at org.eclipse.jetty.server.ResourceService.passConditionalHeaders(ResourceService.java:527)
	at org.eclipse.jetty.server.ResourceService.sendWelcome(ResourceService.java:440)
	at org.eclipse.jetty.server.ResourceService.doGet(ResourceService.java:253)
	at org.eclipse.jetty.servlet.DefaultServlet.doGet(DefaultServlet.java:459)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:867)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1623)
	at com.heytherespender.backend.resource.RootResource.doFilter(RootResource.java:35)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:540)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1345)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:480)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1247)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:502)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:364)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:260)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:765)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:683)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

This is because the code treats a `null` base resource differently from an empty array. If the latter was treated the same as `null` we'd get the correct behavior.

**Comment by joakime** [comment]
A `null` Base Resource will break other things in the Servlet spec btw.
It's a good idea to declare one, even if it points to an empty directory.
Or define your own `Resource` implementation.

This will allow the following `ServletContext` methods (used by countless servlet libraries) to work as well:

* `String getRealPath(String path)`
* `URL getResource(String path)`
* `InputStream getResourceAsStream(String path)`
* `Set<String> getResources(String path)`

**Comment by cowwoc** [comment]
@joakime Thanks for the head's up. By "break other things in the Servlet spec" you mean that users will probably not expect to get back `null` from `getRealPath()` so their code will break. Right?

I assume you still plan to fix this NPE?

**Comment by joakime** [comment]
Merged into `jetty-9.4.x` from PR #3326
